File: algomonster/subarray_sum.py
#!/usr/bin/env python3

from collections import defaultdict
import argparse
import sys
import unittest

# tags: prefix sum

# solution: prefix sum + dict
# complexity:
# run-time: O(n)
# space: O(n)
def subarray_sum(arr: list[int], target: int) -> list[int]:
    curr_sum = 0

    # key: prefix_sum
    # value: index
    prefix_sums = defaultdict(int)
    # empty array
    prefix_sums[0] = 0

    for i in range(len(arr)):
        curr_sum += arr[i]
        # sum([i, j]) = sum([0, j]) - sum([0, i-1])
        # 0-----[i<----k---->j]---->
        complement = curr_sum - target

        if complement in prefix_sums:
            return [prefix_sums[complement], i+1]

        # right exclusive index: [)
        prefix_sums[curr_sum] = i + 1

    return []

class TestSubarraySum(unittest.TestCase):
    def test_empty(self):
        self.assertEqual(subarray_sum([], 0), [])

    def test_no_match(self):
        self.assertEqual(subarray_sum([1, 2, 3], 7), [])

    def test_match(self):
        self.assertEqual(subarray_sum([1, 2, 3, 4], 6), [0, 3])

if __name__ == "__main__":
    parser = argparse.ArgumentParser()
    parser.add_argument('--test', action='store_true', help='Run unit tests')
    args = parser.parse_args()

    if args.test:
        sys.exit(unittest.main(argv=[sys.argv[0]]))

    arr = [int(x) for x in input().split()]
    target = int(input())
    res = subarray_sum(arr, target)
    print(" ".join(map(str, res)))

File: algomonster/subarray_sum.cpp
#include <unordered_map>
#include <vector>

// tags: prefix sum

// solution: prefix sum + map
// complexity:
// run-time: O(n)
// space: O(n)
std::vector<int> subarraySum(const std::vector<int>& v, int target)
{
    int currSum = 0;

    // key: prefix sum
    // value: index up to which prefix sum was calculated
    std::unordered_map<int, int> prefixSums;
    // empty array
    prefixSums[0] = 0;

    for (int i = 0; i != v.size(); ++i) {
        currSum += v[i];

        // sum([i, j]) = sum([0, j]) - sum([0, i-1])
        // 0-----[i<--------->j]---->
        // target = currSum - complement
        int complement = currSum - target;

        auto itr = prefixSums.find(complement);
        if (itr != prefixSums.end()) {
            // right-exclusive: [start, end)
            return std::vector<int>{prefixSums[complement], i+1};
        }

        prefixSums[currSum] = i+1;
    }

    return std::vector<int>{};
}